## Predict the Average Rating of Books

In this project, we have worked on a books dataset published by Goodreads. The dataset comprised information/data on the following factors: Book ID, Book title, Authors and Co-Authors, Average rating, ISBN and ISBN 13, Language, Number of pages, Number of ratings count and Text reviews count, Publication date and Publisher.

A book lover before purchasing a book may prefer to consider one or more of the above mentioned factors with average rating, author, publisher being the most important ones.

With the onset of digital revolution, Goodreads as a platform is very popular amongst book readers. They prefer to look at the average rating and also the reviews provided by others.

For book readers like us, alongwith authors, publishers also play an important role. Publishers like Penguin, Harper Collin have always been associated with good books and hence play a very important role in determining the popularity of a book.

The objective behind building the model was to predict the average rating of a book based on various factors which play an important role in the determination.

This model definitely does not judge the books by their cover but by number of pages, language, publisher and author etc.

## Key Findings from the Data

From the Goodreads books dataset, we seek to find out the following:

1. What is average rating for most of the books?
2. In which year maximum books were published?
3. Does Publication date, number of pages, text reviews count and ratings count really have an impact on the rating of the books?
4. Who are the top 20 authors and publishers?
5. In which language are most of the books read?

Let's dive into the dataset to gain insights and build a model to predict average rating.

### 1. Loading Libraries with Installations

In [64]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

### 2. Load the Goodreads Books Dataset.

In [39]:
cols = ['bookID' ,'title',  'authors', 'average_rating', 'isbn', 'isbn13', 'language_code', 'num_pages', 'ratings_count',      
        'text_reviews_count', 'publication_date', 'publisher',  'coauthor']
df = pd.read_csv('books_data.csv', names=cols, skiprows=1,engine='python', quotechar=',', on_bad_lines='warn')

Note, we have used quoting charcter ',' to prevent quoting issues. The engine is python and it is used when number of columns are variable.

### 3. Initial Analysis of the Data

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   bookID              11127 non-null  int64 
 1   title               11127 non-null  object
 2   authors             11127 non-null  object
 3   average_rating      11127 non-null  object
 4   isbn                11127 non-null  object
 5   isbn13              11127 non-null  object
 6   language_code       11127 non-null  object
 7   num_pages           11127 non-null  object
 8   ratings_count       11127 non-null  int64 
 9   text_reviews_count  11127 non-null  int64 
 10  publication_date    11127 non-null  object
 11  publisher           11127 non-null  object
 12  coauthor            4 non-null      object
dtypes: int64(3), object(10)
memory usage: 1.1+ MB


We have analyzed the datatype of respective columns. All are categorical variables, except book ID, ratings count and text reviews count.

In [41]:
# Check the top 5 rows
df.head(5)

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré           4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré           4.49  0439358078  9780439358071   
2                J.K. Rowling           4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré           4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré           4.78  0439682584  9780439682589   

  language_code num_pages  ratings_count  text_reviews_count publication_date  \
0           eng       652        2095690               27591        9/16/2006   
1           eng       870        2153167               29221         9/1/2004   
2           eng       352           6333                 244        11/1/2003   
3           eng       435        2339585               36325         5/1/2004   
4           eng      2690          41428                 164        9/13/2004   

         publisher coauthor  
0  Scholastic Inc.     None  
1  Scholastic Inc.     None  
2       Scholastic     None  
3  Scholastic Inc.     None  
4       Scholastic     None

In [42]:
#Summary Statistics - Non-Categorical
df.describe().applymap('{:,.2f}'.format)

bookID ratings_count text_reviews_count
count  11,127.00     11,127.00          11,127.00
mean   21,310.94     17,936.49             541.86
std    13,093.36    112,479.43           2,576.17
min         1.00          0.00               0.00
25%    10,287.00        104.00               9.00
50%    20,287.00        745.00              47.00
75%    32,104.50      4,993.50             237.50
max    45,641.00  4,597,666.00          94,265.00

Here, we analyze the summary statistics of only Quantitative/Non-Categorical variables

In [43]:
#Summary Statistics - Categorical
df.describe(include='object')

title       authors average_rating   isbn  \
count                    11127         11127          11127  11127   
unique                   10352          6643            213  11126   
top     The Brothers Karamazov  Stephen King           4.00   3.58   
freq                         9            40            219      2   

               isbn13 language_code num_pages publication_date publisher  \
count           11127         11127     11127            11127     11127   
unique          11127            31       999             3682      2295   
top     9780439785969           eng       288        10/1/2005   Vintage   
freq                1          8908       230               56       318   

                        coauthor  
count                          4  
unique                         4  
top     Harvard University Press  
freq                           1

Here, we analyze the summary statistics of only Qualitative/Categorical variables.

The dataset comprises 11127 rows. We observe that no columns have missing data. But, the column named co-author has only 4 rows. In the following section, we will deep dive into the co-author column.

### 4. Data Cleaning

The dataset comprises 13 columns. But we notice, the 13th column 'Co-Author' has only 4 rows whereas all other columns have 11127 rows. Let's examine further.

In [44]:
dirty = df.loc[~df['coauthor'].isnull(), : ]

In [45]:
dirty

bookID                                              title  \
3348   12224  Streetcar Suburbs: The Process of Growth in Bo...   
4702   16914                  The Tolkien Fan's Medieval Reader   
5877   22128                     Patriots (The Coming Collapse)   
8979   34889  Brown's Star Atlas: Showing All The Bright Sta...   

                                       authors  \
3348                           Sam Bass Warner   
4702  David E. Smith (Turgon of TheOneRing.net   
5877                              James Wesley   
8979                                     Brown   

                                         average_rating  isbn      isbn13  \
3348                                  Jr./Sam B. Warner  3.58  0674842111   
4702   one of the founding members of this Tolkien w...  3.58  1593600119   
5877                                             Rawles  3.63  156384155X   
8979                                     Son & Ferguson  0.00  0851742718   

      language_code num_pages  ratings_count  text_reviews_count  \
3348  9780674842113     en-US            236                  61   
4702  9781593600112       eng            400                  26   
5877  9781563841552       eng            342                  38   
8979  9780851742717       eng             49                   0   

     publication_date  publisher                     coauthor  
3348                6  4/20/2004     Harvard University Press  
4702                4   4/6/2004          Cold Spring Press\t  
5877                4  1/15/1999  Huntington House Publishers  
8979                0   5/1/1977    Brown Son & Ferguson Ltd.

The average rating for these four rows is authors' name.

In [46]:
coauthor = dirty.loc[:, 'average_rating']
coauthor = coauthor.rename("author2")

We will shift every other column to the left and we call this right side of data frame. 

In [47]:
right = dirty.loc[:, 'average_rating':'coauthor'].shift(periods=-1, axis=1)

All the other columns before author and taken as left. 

In [48]:
left = dirty.loc[:, :'authors']

We will merge the co author with author.  

In [49]:
left['authors'] = left['authors'] + coauthor

The left side will look like this. 

In [50]:
left

bookID                                              title  \
3348   12224  Streetcar Suburbs: The Process of Growth in Bo...   
4702   16914                  The Tolkien Fan's Medieval Reader   
5877   22128                     Patriots (The Coming Collapse)   
8979   34889  Brown's Star Atlas: Showing All The Bright Sta...   

                                                authors  
3348                  Sam Bass Warner Jr./Sam B. Warner  
4702  David E. Smith (Turgon of TheOneRing.net one o...  
5877                                James Wesley Rawles  
8979                               Brown Son & Ferguson

The Right side

In [51]:
right

average_rating        isbn         isbn13 language_code  num_pages  \
3348           3.58  0674842111  9780674842113         en-US        236   
4702           3.58  1593600119  9781593600112           eng        400   
5877           3.63  156384155X  9781563841552           eng        342   
8979           0.00  0851742718  9780851742717           eng         49   

      ratings_count text_reviews_count publication_date  \
3348             61                  6        4/20/2004   
4702             26                  4         4/6/2004   
5877             38                  4        1/15/1999   
8979              0                  0         5/1/1977   

                        publisher coauthor  
3348     Harvard University Press      NaN  
4702          Cold Spring Press\t      NaN  
5877  Huntington House Publishers      NaN  
8979    Brown Son & Ferguson Ltd.      NaN

We will combine them. and call it clean. 

In [52]:
clean = pd.concat([left,  right, coauthor], axis=1)

In [53]:
clean = clean.drop(columns= ['coauthor'])

In [54]:
clean

bookID                                              title  \
3348   12224  Streetcar Suburbs: The Process of Growth in Bo...   
4702   16914                  The Tolkien Fan's Medieval Reader   
5877   22128                     Patriots (The Coming Collapse)   
8979   34889  Brown's Star Atlas: Showing All The Bright Sta...   

                                                authors average_rating  \
3348                  Sam Bass Warner Jr./Sam B. Warner           3.58   
4702  David E. Smith (Turgon of TheOneRing.net one o...           3.58   
5877                                James Wesley Rawles           3.63   
8979                               Brown Son & Ferguson           0.00   

            isbn         isbn13 language_code  num_pages  ratings_count  \
3348  0674842111  9780674842113         en-US        236             61   
4702  1593600119  9781593600112           eng        400             26   
5877  156384155X  9781563841552           eng        342             38   
8979  0851742718  9780851742717           eng         49              0   

     text_reviews_count publication_date                    publisher  \
3348                  6        4/20/2004     Harvard University Press   
4702                  4         4/6/2004          Cold Spring Press\t   
5877                  4        1/15/1999  Huntington House Publishers   
8979                  0         5/1/1977    Brown Son & Ferguson Ltd.   

                                                author2  
3348                                  Jr./Sam B. Warner  
4702   one of the founding members of this Tolkien w...  
5877                                             Rawles  
8979                                     Son & Ferguson

We will replace the dirty rows with this newly created clean data frame. 

In [55]:
df.loc[clean.index, :] = clean

While examining the dataset, we found out that for the 4 rows under dicussion, there was a shift of data by 1 column. This means that some of the authors had shifted to the 'Average Rating' column. The average ratings for the respective rows had shifted to the 'ISBN' column and so forth. Hence, we had a 13th column which had publishers' name due to this data shift.

We have corrected this above and the dataframe 'df' is now clean.

### 5. Removing the Columns

We will remove 'coauthor', 'bookID', 'isbn', 'isbn13' as we will not consider them in building the Average Rating Prediction Model.

In [57]:
df = df.drop(columns= ['coauthor', 'bookID', 'isbn', 'isbn13'])

In [58]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   title               11127 non-null  object
 1   authors             11127 non-null  object
 2   average_rating      11127 non-null  object
 3   language_code       11127 non-null  object
 4   num_pages           11127 non-null  object
 5   ratings_count       11127 non-null  int64 
 6   text_reviews_count  11127 non-null  object
 7   publication_date    11127 non-null  object
 8   publisher           11127 non-null  object
dtypes: int64(1), object(8)
memory usage: 782.5+ KB


### 6. Changing the Data type of Variables

After analysing, we noticed that the datatype of the quantitative/non-categorical variables such as average average_rating, num_pages, text_reviews_count is object. The publication_date which is a date is also object.

For building a datset suitable for predictive modelling, we will change the data type of the variables to numeric.

In [60]:
#Changing the data type of the variables to Float/Int
df['average_rating'] = df['average_rating'].astype(np.float32)
df['num_pages'] = df['num_pages'].astype(np.int32)
df['ratings_count'] = df['ratings_count'].astype(np.int32)
df['text_reviews_count'] = df['text_reviews_count'].astype(np.int32)

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   title               11127 non-null  object 
 1   authors             11127 non-null  object 
 2   average_rating      11127 non-null  float32
 3   language_code       11127 non-null  object 
 4   num_pages           11127 non-null  int32  
 5   ratings_count       11127 non-null  int32  
 6   text_reviews_count  11127 non-null  int32  
 7   publication_date    11127 non-null  object 
 8   publisher           11127 non-null  object 
dtypes: float32(1), int32(3), object(5)
memory usage: 608.6+ KB


Checking if the data is clean and there are no unwanted data and columns

In [62]:
df.iloc[3348, :]

title                 Streetcar Suburbs: The Process of Growth in Bo...
authors                               Sam Bass Warner Jr./Sam B. Warner
average_rating                                                     3.58
language_code                                                     en-US
num_pages                                                           236
ratings_count                                                        61
text_reviews_count                                                    6
publication_date                                              4/20/2004
publisher                                      Harvard University Press
Name: 3348, dtype: object

### 7. Converting Dates

In this section, we convert the data type of publication date to datetime as the name of the variable suggests. However, during analysis, it was detected that 2 dates were unfeasible. For example: November and June had 31 days instead of 30. We corrected the respective dates and assumed that it will be the 30th day of the month under consideration.